File: examples/hello_world.ts
import { loadPlugin, importFromPlugin, importAsyncFromPlugin } from 'file:///home/srikanth/rust/calcite-ts/calcite.ts';
///CHANGE TO DENO>LAND OR DENOPKG!!!!

await loadPlugin("hello_world", "file://target/debug/examples/")

const test = importFromPlugin("test");

let arr = new Int32Array([1, 4, 7]);
console.log(`Got result ${test(["hello", "hi"], { num: 8, name: "abc" }, arr)} from plugin`)

console.log(`Value of arr after function call is ${arr}`)

const async_test = importAsyncFromPlugin("async_test");

async_test("hey sleep for 5 seconds", 5).then((reply) => console.log(`Got ${reply}`)).catch((err)=>console.log(`Got err ${err}`));
async_test("hey sleep for 3 seconds", 3).then((reply) => console.log(`Got ${reply}`)).catch((err)=>console.log(`Got err ${err}`));

File: examples/hello_world.rs
//CREATE A PROPER EXAMPLE
#[macro_use]
extern crate serde_derive;

#[derive(Debug, Deserialize)]
struct Test<'a> {
    num: i32,
    name: &'a str,
}

#[calcite::deno_op]
fn test(a: Vec<&str>, b: Test, mut c: calcite::ArrayBuffer<i32>) -> i32 {
    println!("Got first argument {:?}", a);
    println!("Got second argument {:?}", b);
    println!("Got third argument {:?}", c.as_slice());
    c.as_mut_slice()[0] = 45;
    8
}

#[calcite::deno_op]
async fn async_test(msg: &str, secs: u64) -> Result<u64, String> {
    let (tx, rx) = futures::channel::oneshot::channel();
    std::thread::spawn(move || {
        std::thread::sleep(std::time::Duration::from_secs(secs));
        tx.send(()).unwrap();
    });
    rx.await.unwrap();
    println!("Got message {}", msg);
    Err(format!("ERROR SLEPT FOR {} seconds", secs))
}

#[calcite::deno_op]
fn foo() -> Result<String, String> {
    Err("heya".into())
}

calcite::export!(test, async_test, foo);
